### PR TITLE
dbft: always clear BlockQueue tasks after new block is accepted

### PR DESCRIPTION
**Pull request by AnnaShaleva**
It's possible that new block is received via P2P before dBFT finishes the consensus process for this height. On networks with large delays it just means that some other CN collected all commits earlier than other CNs and it relays the accepted block. This situation is normal and in this case we need to store the accepted block and initialize consensus at the next height, and we properly perform these actions. However, if block queue's tasks are not cleared for the accepted height, then the node ends up in a situation when WithVerifyPrepareRequest callback fails to insert updated parent for the next height's PrepareRequest.

WithVerifyPrepareRequest puts parent to the BlockQueue, BlockQueue finds the corresponding miner's task for it and pushes block into miner's channel. However, miner refused to store this block since it has removed all stale tasks earlier for this height.

E.g. in the node logs we firstly see that block 172 was received via P2P before the dBFT able to collect all Commits. After that dBFT is initialized at the next height (173):
```
    2024-01-24T10:54:26.545Z        INFO    dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507     received Commit {"validator": 3}
    2024-01-24T10:54:26.545Z        DEBUG   dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57     not enough to commit    {"count": 4}
    INFO [01-24|10:54:26.963] Imported new chain segment               number=172 hash=70b13e..d272b7 blocks=1 txs=0 mgas=0.000 elapsed=5.205ms     mgasps=0.000 triedirty=0.00B
    INFO [01-24|10:54:26.964] New block in the chain                   "dbft index"=172 "chain index"=172 hash=0x70b13e7566ef532085dcaa779c287151f29f0d17d85cf5b5367d5eba51d272b7 "parent hash"=0xdc644e1c12f877487858e0e626ff73b7d0c663926a39fccf07ac14a273cc9158 primary=4 coinbase=0x1212100000000000000000000000000000000003
    INFO [01-24|10:54:26.964] Fetching latest sealing proposal         "desired number"=173
    INFO [01-24|10:54:26.964] Commit new sealing work                  number=173 sealhash=355652..67baba etherbase=0x1212100000000000000000000000000000000003 txs=0 gas=0 fees=0 elapsed="100.499µs"
    INFO [01-24|10:54:27.364] Imported new chain segment               number=172 hash=4bf439..c9f324 blocks=1 txs=0 mgas=0.000 elapsed=3.235ms     mgasps=0.000 triedirty=0.00B
    INFO [01-24|10:54:27.964] Sealing proposal updated                 number=173 sealhash=355652..67baba "parent hash"=70b13e..d272b7 txs=0
    2024-01-24T10:54:27.964Z        INFO    dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107     initializing dbft       {"height": 173, "view": 0, "index": 6, "role": "Backup"}
 ```

However, BlockQueue still keeping sealing task for height 172. Once PrepareRequest for height 173 arrives, WithVerifyPrepareRequest callback sends the updated parent to the BlockQueue which sends parent to the miner. And miner logs error in a while:
```
    2024-01-24T10:54:37.165Z        DEBUG   dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214     received message        {"type": "PrepareRequest", "from": 5, "height": 173, "view": 0, "my_height": 173, "my_view": 0}
    2024-01-24T10:54:37.166Z        DEBUG   dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:325     received empty PrepareRequest
    2024-01-24T10:54:37.166Z        INFO    dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:333     received PrepareRequest {"validator": 5, "tx": 0}
    2024-01-24T10:54:37.166Z        INFO    dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:106     sending PrepareResponse {"height": 173, "view": 0}
    2024-01-24T10:54:37.166Z        DEBUG   dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9       broadcasting message    {"type": "PrepareResponse", "height": 173, "view": 0}
    ERROR[01-24|10:54:37.166] Block found but no relative pending task number=172 sealhash=00e5e9..661cff hash=888867..8c5916
```

Node continues consensus process and once it accepts 173 block, the following error occurs on attempt to store 173 block (parent block 172 is unknown):
```
2024-01-24T10:54:37.776Z	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 5, "height": 173, "view": 0, "my_height": 173, "my_view": 0}
2024-01-24T10:54:37.777Z	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 5}
2024-01-24T10:54:37.777Z	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 173, "hash": "27cd4cb41d22fdb363c9d99d3ddd519551b61f9746e196cc030216b0887be85d", "tx_count": 0, "merkle": "4e268cc3879dced285c623f0cc8105e2343336810e46823e926bd6edd995a7b1", "prev": "888867dfe4ae19a79d6e807b4a978ac3f394157619b2f71e3f9682ee798c5916"}
WARN [01-24|10:54:37.777] Failed to load old bad blocks            error="pebble: not found"
ERROR[01-24|10:54:37.780] 
########## BAD BLOCK #########
Block: 173 (0xc044ae15313c9fab23baef365fcf5673f16ef330f90a8de5532f104a7a7c31d1)
Error: unknown ancestor
Platform: geth (devel) go1.21.5 amd64 linux
VCS: e9d041e7-20240123
Chain config: &params.ChainConfig{ChainID:12227330, HomesteadBlock:0, DAOForkBlock:<nil>, DAOForkSupport:false, EIP150Block:0, EIP155Block:0, EIP158Block:0, ByzantiumBlock:0, ConstantinopleBlock:0, PetersburgBlock:0, IstanbulBlock:0, MuirGlacierBlock:0, BerlinBlock:0, LondonBlock:0, ArrowGlacierBlock:0, GrayGlacierBlock:0, MergeNetsplitBlock:<nil>, ShanghaiTime:(*uint64)(nil), CancunTime:(*uint64)(nil), PragueTime:(*uint64)(nil), VerkleTime:(*uint64)(nil), TerminalTotalDifficulty:<nil>, TerminalTotalDifficultyPassed:false, Ethash:(*params.EthashConfig)(nil), Clique:(*params.CliqueConfig)(nil), DBFT:(*params.DBFTConfig)(0xc00036e100), IsDevMode:false}
Receipts: 
##############################
 
WARN [01-24|10:54:37.780] error on enqueue block                   error="failed to insert block into chain: unknown ancestor"
```

### CODE REVIEW

**Comment by txhsl** [comment]
I notice there was a view changing happened on 4 of 7 CNs at height 1419, but the block was approved. Is it relevant?

**Comment by AnnaShaleva** [comment]
> I notice there was a view changing happened on 4 of 7 CNs at height 1419

From what I see, non of CNs have sent CV messages for this height and none of them requested CV. They all were at view 0, so no CV happened. If you're talking about `ignoring PrepareRequest due to view changing` messages, it's just an inaccurate log message from the dBFT side, check the following `"viewChanging": false` log. I'll fix this log message soon on the dBFT library side, ref. https://github.com/bane-labs/go-ethereum/issues/106.

Could you provide the exact CV logs (are we talking about `log0125` logs)?

**Comment by txhsl** [comment]
> From what I see, non of CNs have sent CV messages for this height and none of them requested CV. They all were at view 0, so no CV happened.

Yep, logs are a bit strange, but it's good to hear that no CV happened.

> Could you provide the exact CV logs (are we talking about `log0125` logs)?

Yep, we are talking about the same one.

File: miner/worker.go [hunk 1]
```diff
@@ -658,6 +658,7 @@ func (w *worker) resultLoop() {
 			w.pendingMu.RUnlock()
 			if !exist {
 				log.Error("Block found but no relative pending task", "number", block.Number(), "sealhash", sealhash, "hash", hash)
+
```
Comment (roman-khimov) [inline miner/worker.go]: :eyes:

Comment (AnnaShaleva) [reply miner/worker.go]: Fixed!

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**